### DISCUSSION TITLE
Replacing IPFS as the underlying layer

### DISCUSSION THREAD

**Discussion by pstashp** [wnfs-wg/spec#77] [General]
In the current form, is it feasible to replace IPFS as the storage layer, with another ID-addressable platform?

**Comment by matheus23** [comment]
What are you trying to achieve @pstashp? Your question is quite broad and I feel like perhaps I'd be better at answering a more specific question.

I'll try anyways.

In terms of programs, rs-wnfs is not dependent on IPFS implementations. `rs-wnfs` is merely a library that you need to hook up with something else that actually stores the data.

But it builds upon IPFS standards including [IPLD](https://ipld.io/), [CID](https://docs.ipfs.tech/concepts/content-addressing/#identifier-formats)s, UnixFS and more, which make it exceptionally good at *being compatible* with existing IPFS implementations.
E.g. you can use kubo to store rs-wnfs and run `ipfs pin` on another machine to replicate a WNFS.
Or you can use `ipfs get` in kubo with a public WNFS' file CID to download that public file into your local file system.

However, whatever you hook `rs-wnfs` up with as a storage layer only needs to know about CIDs, because it needs to provide a blockstore implementation. People have already built [blockstore adapters on top of amazon S3](https://github.com/ipfs/js-stores/tree/main/packages/datastore-s3) or LevelDB, Web APIs like IndexedDB or native file systems.

**Reply by pstashp** [reply]
Thank you. I was indeed looking to see how easy it would be replace the datastore layer with something else like another blockchain (immutable) or S3 (mutable).  With the sample adapter and your response, I think the answer is a clear yes.

Previously, I wasn't sure about that due to the reliance of IPLD (which made me think that any effort to replace the datastore from IPFS to something else would just be "in theory, it's possible, but some effort is needed")

**Reply by matheus23** [reply]
In principle, implementing the BlockStore doesn't require you to understand IPLD, however, we specifically make use of IPLD's property to list links to further blocks (CIDs) given a block and it's CID. This allows you to walk the whole DAG in order to package up state snapshots or diffs to send to other peers.

**Reply by DougAnderson444** [reply]
@matheus23 Philipp has anyone published their `Blockstore` impl over the Web APIs though?

**Reply by matheus23** [reply]
@DougAnderson444 there's now a helia BlockStore adapter for rs-wnfs:
https://github.com/shovelers/helia-wnfs

You can use that to hook up indexeddb to wnfs.

There's also the nest package: https://github.com/wnfs-wg/nest/ which also exposes an indexeddb based BlockStore IIRC.

**Reply by DougAnderson444** [reply]
Ok! Since those are in JavaScript, I might as well link to my wasm-bindgen version that can be used directly from Rust:

https://github.com/PeerPiper/peerpiper/blob/04847dd5ff96cd8d8ee0c4dbdb435830490be3fe/crates/peerpiper-browser/src/bindgen/blockstore_idb.rs#L121

This also binds to Helia's `blockstore-idb` but does it from Rust
